File: app/src/main/java/com/czbix/v2ex/util/MiscUtils.java
package com.czbix.v2ex.util;


import android.app.Activity;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.support.annotation.Nullable;
import android.support.customtabs.CustomTabsIntent;
import android.widget.Toast;

import com.czbix.v2ex.AppCtx;
import com.czbix.v2ex.BuildConfig;
import com.czbix.v2ex.R;
import com.czbix.v2ex.helper.CustomTabsHelper;
import com.czbix.v2ex.network.RequestHelper;

import java.util.List;

import static android.os.Build.VERSION;
import static android.os.Build.VERSION_CODES;

public class MiscUtils {
    private static final String HOST_MASTER;
    private static final String HOST_WWW;
    private static final String PREFIX_TOPIC;
    private static final String PREFIX_NODE;
    public static final String PREFIX_MEMBER;

    public static final boolean HAS_L;
    public static final boolean HAS_JB;

    static {
        final AppCtx context = AppCtx.getInstance();
        HOST_MASTER = context.getString(R.string.master_host);
        HOST_WWW = context.getString(R.string.www_host);

        PREFIX_TOPIC = context.getString(R.string.topic_url_prefix);
        PREFIX_NODE = context.getString(R.string.node_url_prefix);
        PREFIX_MEMBER = context.getString(R.string.member_url_prefix);

        final int sdkInt = VERSION.SDK_INT;
        HAS_L = sdkInt >= VERSION_CODES.LOLLIPOP;
        HAS_JB = sdkInt >= VERSION_CODES.JELLY_BEAN;
    }

    public static void setClipboard(Context context,@Nullable String title, String str) {
        final ClipboardManager manager = (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        manager.setPrimaryClip(ClipData.newPlainText(title, str));

        Toast.makeText(context, R.string.toast_copied, Toast.LENGTH_SHORT).show();
    }

    public static boolean isEmailLink(String url) {
        return url.startsWith("mailto:");
    }

    public static Intent getEmailIntent(String url) {
        Uri uri = Uri.parse(url);
        return new Intent(Intent.ACTION_SENDTO, uri);
    }

    public static Uri formatUri(String url) {
        Uri uri = Uri.parse(url);

        if (uri.isRelative()) {
            uri = Uri.parse(RequestHelper.BASE_URL + url);
        }

        return uri;
    }

    public static void openUrl(Activity activity, String url) {
        if (isEmailLink(url)) {
            final Intent emailIntent = getEmailIntent(url);
            activity.startActivity(emailIntent);
            return;
        }

        final Uri uri = formatUri(url);
        final Intent intent = new Intent(Intent.ACTION_VIEW, uri);
        final PackageManager packageManager = activity.getPackageManager();
        final List<ResolveInfo> resolveInfos = packageManager.queryIntentActivities(intent, 0);

        boolean findMyself = false;
        for (ResolveInfo resolveInfo : resolveInfos) {
            if (resolveInfo.activityInfo.packageName.equals(BuildConfig.APPLICATION_ID)) {
                findMyself = true;
                break;
            }
        }

        if (findMyself) {
            intent.setPackage(BuildConfig.APPLICATION_ID);
            activity.startActivity(intent);
            return;
        }

        final CustomTabsIntent.Builder builder = CustomTabsHelper.getBuilder(activity, null);
        builder.build().launchUrl(activity, uri);
    }
}

File: app/src/main/kotlin/com/czbix/v2ex/util/MiscUtils.kt
package com.czbix.v2ex.util


import android.app.Activity
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.res.Resources
import android.net.Uri
import android.os.Build
import android.widget.Toast
import androidx.core.os.ConfigurationCompat
import com.czbix.v2ex.BuildConfig
import com.czbix.v2ex.R
import com.czbix.v2ex.helper.CustomTabsHelper
import com.czbix.v2ex.network.RequestHelper

object MiscUtils {
    @JvmStatic
    val appUpdateIntent = Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=${BuildConfig.APPLICATION_ID}"))

    @JvmStatic
    fun setClipboard(context: Context, data: ClipData) {
        val manager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        manager.setPrimaryClip(data)

        Toast.makeText(context, R.string.toast_copied, Toast.LENGTH_SHORT).show()
    }

    @JvmStatic
    fun setClipboard(context: Context, title: String?, str: String) {
        setClipboard(context, ClipData.newPlainText(title, str))
    }

    fun isEmailLink(url: String): Boolean {
        return url.startsWith("mailto:")
    }

    fun getEmailIntent(url: String): Intent {
        val uri = Uri.parse(url)
        return Intent(Intent.ACTION_SENDTO, uri)
    }

    @JvmStatic
    fun formatUrl(url: String): String {
        return when {
            url.startsWith("//") -> "https:$url"
            url.startsWith("/") -> RequestHelper.BASE_URL + url
            else -> url
        }
    }

    @JvmStatic
    @JvmOverloads
    fun openUrl(activity: Activity, url: String, findMyself: Boolean = true) {
        if (isEmailLink(url)) {
            val emailIntent = getEmailIntent(url)
            activity.startActivity(emailIntent)
            return
        }

        val uri = Uri.parse(formatUrl(url))
        val intent = Intent(Intent.ACTION_VIEW, uri)
        val packageManager = activity.packageManager
        val resolveInfos = packageManager.queryIntentActivities(intent, 0)

        if (findMyself) {
            val foundMyself = resolveInfos.any {
                it.activityInfo.packageName == BuildConfig.APPLICATION_ID
            }

            if (foundMyself) {
                intent.`package` = BuildConfig.APPLICATION_ID
                activity.startActivity(intent)
                return
            }
        }


        val builder = CustomTabsHelper.getBuilder(activity, null)
        builder.build().launchUrl(activity, uri)
    }

    @JvmStatic
    fun getViewImageIntent(context: Context, uri: Uri): Intent {
        val googlePhotoPackage = "com.google.android.apps.photos"

        val intent = Intent()
        intent.action = Intent.ACTION_VIEW
        intent.setDataAndTypeAndNormalize(uri, "image/*")
        intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)

        val packageManager = context.packageManager
        val resolveInfos = packageManager.queryIntentActivities(intent, 0)
        val foundMyself = resolveInfos.any {
            it.activityInfo.packageName == googlePhotoPackage
        }

        if (foundMyself) {
            intent.`package` = googlePhotoPackage
        }

        return intent
    }

    fun isChineseLang(resources: Resources): Boolean {
        val lang = ConfigurationCompat.getLocales(resources.configuration)[0]!!.language
        return lang == "zh"
    }

    fun atLeast(version: Int): Boolean {
        return Build.VERSION.SDK_INT >= version
    }

    @JvmStatic
    fun decodeCfEmail(str: String): String {
        val sb = StringBuilder()
        val k = str.substring(0, 2).toInt(16)
        for (i in 2.until(str.length).step(2)) {
            val n = str.substring(i, i + 2).toInt(16) xor k
            sb.append(n.toChar())
        }

        return sb.toString()
    }
}

inline fun <reified T : Any> getLogTag(): String {
    return T::class.java.simpleName
}